File: Source/Extensions/Foundation/NSString+OFExtension.m
#import "NSString+OFExtension.h"

@implementation NSString (OFExtension)

#pragma mark - Reading Time

- (NSTimeInterval)of_readingTime {
  return [self of_readingTimeWithSpeed:200];
}

- (NSTimeInterval)of_readingTimeWithSpeed:(NSUInteger)speed {
  NSMutableArray *words = [[self componentsSeparatedByCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]] mutableCopy];
  [words removeObject:@""];
  return words.count / (speed * 60.0);
}

#pragma mark - Regex

- (BOOL)of_matchRegex:(NSString *)pattern {
  return [[NSPredicate predicateWithFormat:@"SELF MATCHES %@", pattern] evaluateWithObject:self];
}

- (NSString *)of_firstMatchOfRegex:(NSString *)pattern {
  NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:pattern options:0 error:NULL];
  NSTextCheckingResult *match = [regex firstMatchInString:self options:0 range:NSMakeRange(0, self.length)];
  if (match == nil) {
    return nil;
  }
  return [self substringWithRange:match.range];
}

- (NSArray<NSString *> *)of_allMatchesOfRegex:(NSString *)pattern {
  NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:pattern options:0 error:NULL];
  NSArray *matches = [regex matchesInString:self options:0 range:NSMakeRange(0, self.length)];
  if (matches == nil) {
    return @[];
  }
  NSMutableArray *substrings = [[NSMutableArray alloc] initWithCapacity:matches.count];
  for (NSTextCheckingResult *match in matches) {
    [substrings addObject:[self substringWithRange:match.range]];
  }
  return [substrings copy];
}

- (BOOL)of_matchEmailString {
  return [self of_matchRegex:@"^[_A-Za-z0-9-+]+(\\.[_A-Za-z0-9-+]+)*@[A-Za-z0-9-]+(\\.[A-Za-z0-9-]+)*(\\.[A-Za-z‌​]{2,})$"];
}

- (BOOL)of_matchUrlString {
  return [NSURL URLWithString:self] != nil;
}

@end

File: Source/Extensions/Foundation/NSString+OFExtension.swift
import Foundation

public extension String {
  
  // MARK: - To Number
  
  public func of_toInt() -> Int? {
    return NSNumberFormatter().numberFromString(self)?.integerValue
  }
  
  public func of_toDouble() -> Double? {
    return NSNumberFormatter().numberFromString(self)?.doubleValue
  }
  
  // MARK: - Reading Time
  
  // speed = words per minute. default speed is 200
  public func of_readingTime(speed speed: UInt = 200) -> NSTimeInterval {
    return (self as NSString).of_readingTimeWithSpeed(speed)
  }
  
  // MARK: - Regex
  
  public func of_matchRegex(pattern: String) -> Bool {
    return (self as NSString).of_matchRegex(pattern)
  }
  
  public func of_firstMatchOfRegex(pattern: String) -> String? {
    return (self as NSString).of_firstMatchOfRegex(pattern)
  }
  
  public func of_allMatchesOfRegex(pattern: String) -> [String] {
    return (self as NSString).of_allMatchesOfRegex(pattern)
  }
  
  public func of_matchEmailString() -> Bool {
    return (self as NSString).of_matchEmailString()
  }
  
  public func of_matchUrlString() -> Bool {
    return (self as NSString).of_matchUrlString()
  }
  
}